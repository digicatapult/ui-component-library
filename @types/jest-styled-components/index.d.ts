declare module 'jest-styled-components/serializer'
