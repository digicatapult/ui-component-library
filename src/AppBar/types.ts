import React, { PropsWithChildren } from 'react'

// TODO create css type, react might have it // break it down
export interface Props {
  shadow?: boolean
  active?: boolean
  fixed?: boolean
  color?: string
  width?: string
  direction?: 'row' | 'column'
  theme?: {
    [key: string]: string
  }
}

export interface IAppBar {
  (args: PropsWithChildren<Props> & Props): React.ReactElement
  Item: React.FC<PropsWithChildren<Props>>
}

export interface IItem {
  (args: PropsWithChildren<Props> & Props): React.ReactElement
}
