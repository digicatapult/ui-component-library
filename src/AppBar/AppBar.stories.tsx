import React from 'react'
import { Story, Meta } from '@storybook/react'

import AppBar from './index'

export default {
  title: 'Components/AppBar',
  component: AppBar,
  parameters: {
    layout: 'fullscreen',
  },
  argTypes: {
    fixed: {
      control: {
        type: 'boolean',
      },
    },
    width: {
      control: {
        type: 'string',
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['default', 'inverted'],
      },
    },
  },
  args: {
    fixed: false,
    variant: 'default',
    width: '100%',
    children: [<div>1</div>, <div>2</div>],
  },
} as Meta<typeof AppBar>

export const Playground: Story<typeof AppBar> = (args) => <AppBar {...args} />
export const Default: Story<typeof AppBar> = () => <AppBar />
export const DefaultWithChildren: Story<typeof AppBar> = () => (
  <AppBar>
    <div>1</div>
    <div>2</div>
    <div>3</div>
  </AppBar>
)
