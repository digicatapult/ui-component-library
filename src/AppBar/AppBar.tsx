import React, { PropsWithChildren } from 'react'
import styled from 'styled-components'

// TODO create css type, react might have it
interface AppBarProps {
  fixed?: boolean
  search?: boolean
  shadow?: boolean
  width?: string
  padding?: string
  direction?: string
  position?: string
  background?: string
  theme?: object
}

export interface AppBar extends React.FC<PropsWithChildren<AppBarProps>> {
  Item?: React.FC<PropsWithChildren<AppBarProps>>
}

const Theme = {
  primary: '#FFA',
  accent: '#000',
}

// TODO props per dom element as well as all for appbar
const Wrapper = styled('div')<AppBarProps>`
  display: flex;
  width: ${({ width }) => width || '100%'};
  flex-direction: row;
  justify-content: space-between;
  box-sizing: border-box;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: ${({ shadow }) =>
    shadow
      ? '0px 2px 4px -1px rgb(0 0 0 / 20%), 0px 4px 5px 0px rgb(0 0 0 / 14%), 0px 1px 10px 0px rgb(0 0 0 / 12%);'
      : 'none'};
  background-color: ${({ color, theme }) => color || theme.primary};
  color: ${({ theme }) => theme.accent || '#000'};
`

const Search = styled('div')<AppBarProps>`
  position: relative;
  width: 75px;
  hegith: 100%;
`

const Bar = styled('div')<AppBarProps>`
  position: ${({ fixed }) => (fixed ? 'fixed' : 'relative')};
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  height: 100%;
  min-height: 74px;
  gap: 15px;
  flex-direction: ${({ direction }) => direction || 'row'};
  padding: ${({ padding }) => padding || '10px 25px'};
  padding-left: 54px;
`

const ItemStyle = styled('div')<AppBarProps>`
  font-family: Roboto;
  font-style: normal;
  font-weight: 700;
  padding: 0px 10px;
  text-transform: uppercase;
  font-size: 16px;
  line-height: 15px;
`

const Item: React.FC<PropsWithChildren<AppBarProps>> = ({
  children,
}: PropsWithChildren<AppBarProps>) => {
  return <ItemStyle>{children}</ItemStyle>
}

// TODO add support for logo icon? - not needed for this project
const AppBar: AppBar = ({
  children,
  search = false,
  shadow = false,
  theme = Theme,
  ...props
}) => {
  return (
    <Wrapper {...props} theme={theme} shadow={shadow}>
      <Bar>{children}</Bar>
      {search && <Search {...props} />}
    </Wrapper>
  )
}

AppBar.Item = Item

export default AppBar
