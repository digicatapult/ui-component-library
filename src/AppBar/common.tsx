import styled from 'styled-components'
import { Props } from './types.js'

// TODO props per dom element as well as all for appbar
// TODO abstract into small component for reusing?
const Wrapper = styled('div')`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  box-sizing: border-box;
  transition: box-shadow 1s cubic-bezier(0.8, 1, 0.2, 1) 3s;
  ${({ shadow, width, color, theme }: Props) => `
    width: ${width || '100%'};
    box-shadow: ${
      shadow || false
        ? '0px 3px 5px -2px rgb(0 0 0 / 20%), 0px 4px 5px 0px rgb(0 0 0 / 14%), 0px 1px 10px 0px rgb(0 0 0 / 12%);'
        : 'none'
    };
    background-color: ${color || theme?.primary || '#FFFE'};
    color: ${theme?.accent || '#000'};
  `}
`

const ToolBar = styled('ul')`
  ${({ fixed, direction }: Props) => `
    position: ${fixed ? 'fixed' : 'relative'};
    flex-direction: ${direction || 'row'};
  `}
  padding: 0px 25px;
  margin: 0;
  list-style-type: none;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  height: 100%;
  min-height: 74px;
  gap: 10px;
  padding-left: 54px;
`

const Link = styled('a')`
  display: inline-block;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  color: inherit;
  font-family: Roboto;
  font-weight: 700;
  padding: 0 15px;
  font-size: 1em;
  line-height: 74px;

  ${({ active, theme }: Props) =>
    active
      ? `
      margin-top: 4px;
      background: linear-gradient(360deg, #DFE667 -408.97%, rgba(223, 230, 103, 0) 78.21%);
      border-bottom: 4px solid ${theme?.accent || '#000'}
    `
      : ``}
`

const Li = styled('li')`
  display: block;
  height: 100%;
  margin: 0;
  padding: 0;
  cursor: pointer;
  transition: all 0.5s;

  &:hover {
    background-color: rgba(40, 30, 30, 0.1);
  }
`

export { Li, ToolBar, Wrapper, Link }
