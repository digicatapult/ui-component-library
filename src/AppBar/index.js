import AppBar from './AppBar.tsx'
export default AppBar
