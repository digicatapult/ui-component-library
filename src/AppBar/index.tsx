import React from 'react'
import { IAppBar, IItem } from './types.js'
import { Li, Link, ToolBar, Wrapper } from './common.js'

const Item: IItem = ({ children, ...props }) => (
  <Li>
    <Link href="#" {...props}>
      {children}
    </Link>
  </Li>
)

// TODO update to use grid comopnent
// TODO move to separate components in case this could be reused
const AppBar: IAppBar = ({ children, ...props }) => (
  <Wrapper {...props}>
    <ToolBar>
      {React.Children.map(children, (child: any) => {
        if (child.props) return React.cloneElement(child, props)
      })?.filter(Boolean)}
    </ToolBar>
  </Wrapper>
)

AppBar.Item = Item

export default AppBar
