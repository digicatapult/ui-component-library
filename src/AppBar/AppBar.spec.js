import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import AppBar from './index.tsx'

describe('AppBar', () => {
  test('default', () => {
    const tree = renderer
      .create(
        <AppBar>
          <AppBar.Item active={true}>what we do</AppBar.Item>
          <AppBar.Item>maps</AppBar.Item>
          <AppBar.Item>contact us</AppBar.Item>
        </AppBar>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('default - fixed', () => {
    const tree = renderer
      .create(
        <AppBar shadow={false} width={'50%'} fixed={true}>
          <AppBar.Item>what we do</AppBar.Item>
          <AppBar.Item>maps</AppBar.Item>
          <AppBar.Item active={true}>contact us</AppBar.Item>
        </AppBar>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('default - custom width', () => {
    const tree = renderer
      .create(
        <AppBar shadow={false} width={'50%'}>
          <AppBar.Item>what we do</AppBar.Item>
          <AppBar.Item>maps</AppBar.Item>
          <AppBar.Item active={true}>contact us</AppBar.Item>
        </AppBar>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('default - no shadow', () => {
    const tree = renderer
      .create(
        <AppBar shadow={false}>
          <AppBar.Item>what we do</AppBar.Item>
          <AppBar.Item>maps</AppBar.Item>
          <AppBar.Item active={true}>contact us</AppBar.Item>
        </AppBar>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('hii', () => {
    const tree = renderer
      .create(
        <AppBar
          shadow={false}
          theme={{
            primary: '#27847A',
            accent: '#FFF',
          }}
        >
          <AppBar.Item>what we do</AppBar.Item>
          <AppBar.Item active={true}>maps</AppBar.Item>
          <AppBar.Item>contact us</AppBar.Item>
        </AppBar>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
