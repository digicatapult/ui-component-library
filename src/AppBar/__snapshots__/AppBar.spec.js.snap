// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  box-sizing: border-box;
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  box-shadow: none;
  background-color: #FFA;
  color: #000;
}

.c1 {
  position: relative;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  height: 100%;
  min-height: 74px;
  gap: 15px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  padding: 10px 25px;
  padding-left: 54px;
}

.c2 {
  font-family: Roboto;
  font-style: normal;
  font-weight: 700;
  padding: 0px 10px;
  text-transform: uppercase;
  font-size: 16px;
  line-height: 15px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      what we do
    </div>
    <div
      className="c2"
    >
      maps
    </div>
    <div
      className="c2"
    >
      contact us
    </div>
  </div>
</div>
`;
